sharpd/sharp_zebra.c

```diff
@@ -351,8 +351,33 @@ void vrf_label_add(vrf_id_t vrf_id, afi_t afi, mpls_label_t label)
 	zclient_send_vrf_label(zclient, vrf_id, afi, label, ZEBRA_LSP_SHARP);
 }
 
-void route_add(const struct prefix *p, vrf_id_t vrf_id,
-	       uint8_t instance, const struct nexthop_group *nhg,
+void nhg_add(uint32_t id, const struct nexthop_group *nhg)
+{
+	struct zapi_nexthop nh_array[MULTIPATH_NUM];
+	struct zapi_nexthop *api_nh;
+	uint16_t nexthop_num = 0;
+	struct nexthop *nh;
+
+	for (ALL_NEXTHOPS_PTR(nhg, nh)) {
+		api_nh = &nh_array[nexthop_num];
```

Comment:
So this is the sort of thing I was concerned about in the comment about lib/zclient: limits check for 'nh_array', something that would have to be maintained in every daemon who wants to use the nhg zapi.

Reply:
Addressed in previous Q. I think its pretty similar to how we are sending routes in sharpd unless I misunderstood your statement.

Reply:
the issue is that this will overrun nh_array if it's not limit-checked. if we're doing something like that in zapi route handling somewhere, it should be fixed too, but it's hard to see adding a new instance of it?

Reply:
ah i see. yes this needs a check.

But I still stand by my other statement related to this that you cant really have a common entry for the converting in to zapi encodings since every daemons idea of a nexthop is different. They are all going to have to encode to zapi_nexthop from their own structs they are using to infer nexthops. I'm pretty sure sharpd is the only one actually using `lib/nexthop_goup`

Reply:
> you cant really have a common entry for the converting in to zapi encodings

I just don't see that - we do it for routes and lsps, for example, among many other objects, and both of those include nexthops? daemons are, as you say, responsible for converting from their own internal datastructs to the neutral zapi objects. but there are a couple of key benefits: there is a single copy of the marshall/unmarshall code, and there is some reasonable way to estimate message buffer/stream size. another benefit is that the zclient api uses the neutral zapi struct - it doesn't have to be updated every time any new field is added to the struct.

Reply:
If we did that we could have a common function for `nexthop_group ->zapi_nexthop`  but every protocol would still need a conversion from `proto independent stuff -> nexthop_group` right? You think that's worth it?

Reply:
I ... guess I think the pattern in use for routes and lsps ought to be the default pattern, unless there's some compelling reason not to follow it. I wouldn't personally suggest trying to use `struct nexthop_group` - that's got a specific meaning for configuration. like the route and lsp zapis, I'm suggesting defining a `zapi_nhg` (or something) struct, and using it in the zclient marshall/unmarshall apis. as you say, `zapi_nhg` should contain some `zapi_nexthop` objects, and we already have code that zapis them, so that's reusable.